File: padlock/src/main/java/com/pyamsoft/padlock/list/LockListFragment.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.list;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter;
import com.pyamsoft.padlock.Injector;
import com.pyamsoft.padlock.PadLock;
import com.pyamsoft.padlock.R;
import com.pyamsoft.padlock.databinding.FragmentLockListBinding;
import com.pyamsoft.padlock.model.AppEntry;
import com.pyamsoft.padlock.onboard.list.OnboardListDialog;
import com.pyamsoft.padlock.pin.PinEntryDialog;
import com.pyamsoft.padlock.service.PadLockService;
import com.pyamsoft.pydroid.design.fab.HideScrollFABBehavior;
import com.pyamsoft.pydroid.design.util.FABUtil;
import com.pyamsoft.pydroid.ui.app.fragment.ActionBarFragment;
import com.pyamsoft.pydroid.ui.loader.ImageLoader;
import com.pyamsoft.pydroid.ui.loader.LoaderHelper;
import com.pyamsoft.pydroid.ui.loader.loaded.Loaded;
import com.pyamsoft.pydroid.ui.rating.RatingDialog;
import com.pyamsoft.pydroid.util.AnimUtil;
import com.pyamsoft.pydroid.util.DialogUtil;
import javax.inject.Inject;
import timber.log.Timber;

public class LockListFragment extends ActionBarFragment {

  @NonNull public static final String TAG = "LockListFragment";
  @SuppressWarnings("WeakerAccess") FastItemAdapter<LockListItem> fastItemAdapter;
  @SuppressWarnings("WeakerAccess") @Inject LockListPresenter presenter;
  @SuppressWarnings("WeakerAccess") FragmentLockListBinding binding;
  @NonNull Loaded fabIconTask = LoaderHelper.empty();
  @NonNull final LockListPresenter.FABStateCallback fabStateCallback =
      new LockListPresenter.FABStateCallback() {

        @Override public void onSetFABStateEnabled() {
          fabIconTask = LoaderHelper.unload(fabIconTask);
          fabIconTask =
              ImageLoader.fromResource(R.drawable.ic_lock_outline_24dp).into(binding.applistFab);
        }

        @Override public void onSetFABStateDisabled() {
          fabIconTask = LoaderHelper.unload(fabIconTask);
          fabIconTask =
              ImageLoader.fromResource(R.drawable.ic_lock_open_24dp).into(binding.applistFab);
        }
      };
  boolean listDoneRefreshing;
  @Nullable private MenuItem displaySystemItem;
  @Nullable private DividerItemDecoration dividerDecoration;
  private FilterListDelegate filterListDelegate;

  @Override public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    listDoneRefreshing = false;
    setHasOptionsMenu(true);

    Injector.get().provideComponent().plusLockListComponent().inject(this);
  }

  @Nullable @Override
  public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    filterListDelegate = new FilterListDelegate();
    fastItemAdapter = new FastItemAdapter<>();
    binding = FragmentLockListBinding.inflate(inflater, container, false);
    return binding.getRoot();
  }

  @Override public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    filterListDelegate.onViewCreated(fastItemAdapter);
    setupRecyclerView();
    setupSwipeRefresh();
    setupFAB();
  }

  @Override public void onStart() {
    super.onStart();

    presenter.registerOnBus(new LockListPresenter.BusCallback() {

      @Override public void onMasterPinCreateSuccess() {
        fabStateCallback.onSetFABStateEnabled();
        final View v = getView();
        if (v != null) {
          Snackbar.make(v, "PadLock Enabled", Snackbar.LENGTH_SHORT).show();
        }
      }

      @Override public void onMasterPinCreateFailure() {
        Toast.makeText(getContext(), "Error: Mismatched PIN", Toast.LENGTH_SHORT).show();
      }

      @Override public void onMasterPinClearSuccess() {
        fabStateCallback.onSetFABStateDisabled();
        final View v = getView();
        if (v != null) {
          Snackbar.make(v, "PadLock Disabled", Snackbar.LENGTH_SHORT).show();
        }
      }

      @Override public void onMasterPinClearFailure() {
        Toast.makeText(getContext(), "Error: Invalid PIN", Toast.LENGTH_SHORT).show();
      }
    });

    presenter.setFABStateFromPreference(fabStateCallback);

    if (!listDoneRefreshing) {
      refreshList(false);
    }
  }

  @Override public void onStop() {
    super.onStop();
    presenter.stop();
  }

  @Override public void onResume() {
    super.onResume();
    setActionBarUpEnabled(false);
    AnimUtil.popShow(binding.applistFab, 300, 400);
  }

  @Override public void onPause() {
    super.onPause();
    AnimUtil.popHide(binding.applistFab, 300, 400);
  }

  private void setupSwipeRefresh() {
    binding.applistSwipeRefresh.setColorSchemeResources(R.color.blue500, R.color.amber700,
        R.color.blue700, R.color.amber500);
    binding.applistSwipeRefresh.setOnRefreshListener(() -> {
      Timber.d("onRefresh");
      refreshList(true);
    });
  }

  private void setupRecyclerView() {
    dividerDecoration = new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL);
    fastItemAdapter.withSelectable(true);
    fastItemAdapter.withOnClickListener((view, iAdapter, item, i) -> {
      displayLockInfoFragment(item.getModel());
      return true;
    });

    LinearLayoutManager manager = new LinearLayoutManager(getContext());
    manager.setItemPrefetchEnabled(true);
    manager.setInitialPrefetchItemCount(3);
    binding.applistRecyclerview.setLayoutManager(manager);
    binding.applistRecyclerview.setClipToPadding(false);
    binding.applistRecyclerview.setHasFixedSize(false);
    binding.applistRecyclerview.addItemDecoration(dividerDecoration);
    binding.applistRecyclerview.setAdapter(fastItemAdapter);
  }

  @Override public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
    super.onCreateOptionsMenu(menu, inflater);
    inflater.inflate(R.menu.locklist_menu, menu);
    inflater.inflate(R.menu.search_menu, menu);
  }

  @Override public void onPrepareOptionsMenu(@NonNull Menu menu) {
    super.onPrepareOptionsMenu(menu);
    setupDisplaySystemVisibleItem(menu);
    filterListDelegate.onPrepareOptionsMenu(menu, fastItemAdapter);
  }

  private void setupDisplaySystemVisibleItem(final @NonNull Menu menu) {
    displaySystemItem = menu.findItem(R.id.menu_is_system);
    presenter.setSystemVisibilityFromPreference(new LockListPresenter.SystemVisibilityCallback() {
      @Override public void onSetSystemVisible() {
        setSystemVisible(true);
      }

      @Override public void onSetSystemInvisible() {
        setSystemVisible(false);
      }
    });
  }

  @SuppressWarnings("WeakerAccess") void setSystemVisible(boolean visible) {
    if (displaySystemItem == null) {
      throw new IllegalStateException("DisplaySystem menu item is NULL.");
    }

    displaySystemItem.setChecked(visible);
  }

  @Override public void onDestroyView() {
    filterListDelegate.onDestroyView();
    displaySystemItem = null;

    binding.applistRecyclerview.removeItemDecoration(dividerDecoration);
    binding.applistRecyclerview.setOnClickListener(null);
    binding.applistRecyclerview.setLayoutManager(null);
    binding.applistRecyclerview.setAdapter(null);

    binding.applistFab.setOnClickListener(null);
    binding.applistSwipeRefresh.setOnRefreshListener(null);

    fabIconTask = LoaderHelper.unload(fabIconTask);
    binding.unbind();
    super.onDestroyView();
  }

  @Override public void onDestroy() {
    super.onDestroy();
    presenter.destroy();
    PadLock.getRefWatcher(this).watch(this);
  }

  @Override public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.menu_is_system:
        if (binding.applistSwipeRefresh != null && !binding.applistSwipeRefresh.isRefreshing()) {
          Timber.d("List is not refreshing. Allow change of system preference");
          if (item.isChecked()) {
            presenter.setSystemInvisible();
          } else {
            presenter.setSystemVisible();
          }

          refreshList(true);
        }
        break;
      default:
        Timber.w("Item selected: %d, do nothing", item.getItemId());
    }
    return super.onOptionsItemSelected(item);
  }

  private void setupFAB() {
    binding.applistFab.setOnClickListener(v -> {
      if (PadLockService.isRunning()) {
        DialogUtil.guaranteeSingleDialogFragment(getActivity(),
            PinEntryDialog.newInstance(getContext().getPackageName(),
                getActivity().getClass().getName()), PinEntryDialog.TAG);
      } else {
        DialogUtil.guaranteeSingleDialogFragment(getActivity(), new AccessibilityRequestDialog(),
            "accessibility");
      }
    });
    FABUtil.setupFABBehavior(binding.applistFab, new HideScrollFABBehavior(24));
  }

  public void onCompletedOnboarding() {
    final FragmentActivity activity = getActivity();
    if (activity instanceof RatingDialog.ChangeLogProvider) {
      RatingDialog.showRatingDialog(activity, (RatingDialog.ChangeLogProvider) activity, false);
    }
  }

  @SuppressWarnings("WeakerAccess") void displayLockInfoFragment(@NonNull AppEntry entry) {
    DialogUtil.guaranteeSingleDialogFragment(getActivity(), LockInfoDialog.newInstance(entry),
        LockInfoDialog.TAG);
  }

  void refreshList(boolean force) {
    fastItemAdapter.clear();
    presenter.populateList(force, new LockListPresenter.PopulateListCallback() {

      private void setRefreshing(boolean refresh) {
        binding.applistSwipeRefresh.post(() -> {
          if (binding != null) {
            if (binding.applistSwipeRefresh != null) {
              binding.applistSwipeRefresh.setRefreshing(refresh);
            }
          }
        });

        final FragmentActivity activity = getActivity();
        if (activity != null) {
          Timber.d("Reload options");
          activity.supportInvalidateOptionsMenu();
        }
      }

      @Override public void onListPopulateBegin() {
        listDoneRefreshing = false;
        setRefreshing(true);
        fastItemAdapter.clear();
        binding.applistFab.hide();
      }

      @Override public void onEntryAddedToList(@NonNull AppEntry entry) {
        fastItemAdapter.add(new LockListItem(entry));
      }

      @Override public void onListPopulated() {
        listDoneRefreshing = true;
        setRefreshing(false);
        binding.applistFab.show();

        if (fastItemAdapter.getAdapterItemCount() > 1) {
          Timber.d("We have refreshed");
          presenter.showOnBoarding(new LockListPresenter.OnboardingCallback() {
            @Override public void onShowOnboarding() {
              Timber.d("Show onboarding");
              DialogUtil.onlyLoadOnceDialogFragment(getActivity(), new OnboardListDialog(),
                  OnboardListDialog.TAG);
            }

            @Override public void onOnboardingComplete() {
              onCompletedOnboarding();
            }
          });
        } else {
          Toast.makeText(getContext(), "Error while loading list. Please try again.",
              Toast.LENGTH_SHORT).show();
        }
      }

      @Override public void onListPopulateError() {
        DialogUtil.guaranteeSingleDialogFragment(getActivity(), new ErrorDialog(), "error");
      }
    });
  }
}

File: padlock/src/main/java/com/pyamsoft/padlock/list/LockListFragment.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.padlock.list

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.pyamsoft.padlock.Injector
import com.pyamsoft.padlock.PadLockComponent
import com.pyamsoft.padlock.R
import com.pyamsoft.padlock.api.service.LockServiceInteractor.ServiceState
import com.pyamsoft.padlock.api.service.LockServiceInteractor.ServiceState.DISABLED
import com.pyamsoft.padlock.api.service.LockServiceInteractor.ServiceState.ENABLED
import com.pyamsoft.padlock.api.service.LockServiceInteractor.ServiceState.PAUSED
import com.pyamsoft.padlock.api.service.LockServiceInteractor.ServiceState.PERMISSION
import com.pyamsoft.padlock.api.service.ServiceManager
import com.pyamsoft.padlock.model.list.AppEntry
import com.pyamsoft.padlock.model.list.ListDiffProvider
import com.pyamsoft.padlock.pin.ClearPinPresenter
import com.pyamsoft.padlock.pin.ConfirmPinPresenter
import com.pyamsoft.padlock.pin.CreatePinPresenter
import com.pyamsoft.padlock.pin.PinConfirmDialog
import com.pyamsoft.padlock.pin.PinCreateDialog
import com.pyamsoft.padlock.service.device.UsagePermissionChecker
import com.pyamsoft.pydroid.core.singleDisposable
import com.pyamsoft.pydroid.core.tryDispose
import com.pyamsoft.pydroid.ui.app.requireToolbarActivity
import com.pyamsoft.pydroid.ui.app.toolbarActivity
import com.pyamsoft.pydroid.ui.util.setUpEnabled
import com.pyamsoft.pydroid.ui.util.show
import timber.log.Timber
import javax.inject.Inject

class LockListFragment : Fragment(),
    CreatePinPresenter.Callback,
    ConfirmPinPresenter.Callback,
    ClearPinPresenter.Callback {

  @field:Inject internal lateinit var viewModel: LockListViewModel
  @field:Inject internal lateinit var serviceManager: ServiceManager
  @field:Inject internal lateinit var lockView: LockListView

  @field:Inject internal lateinit var createPinPresenter: CreatePinPresenter
  @field:Inject internal lateinit var confirmPinPresenter: ConfirmPinPresenter
  @field:Inject internal lateinit var clearPinPresenter: ClearPinPresenter

  private var visibilityDisposable by singleDisposable()
  private var databaseChangeDisposable by singleDisposable()
  private var lockEventDisposable by singleDisposable()
  private var fabStateChangeDisposable by singleDisposable()
  private var fabStateCheckDisposable by singleDisposable()
  private var populateDisposable by singleDisposable()

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    Injector.obtain<PadLockComponent>(requireContext().applicationContext)
        .plusLockListComponent()
        .activity(requireActivity())
        .toolbarActivity(requireToolbarActivity())
        .listStateTag(TAG)
        .owner(viewLifecycleOwner)
        .inflater(inflater)
        .container(container)
        .savedInstanceState(savedInstanceState)
        .diffProvider(object : ListDiffProvider<AppEntry> {
          override fun data(): List<AppEntry> = lockView.getListData()
        })
        .build()
        .inject(this)

    lockView.create()
    return lockView.root()
  }

  override fun onViewCreated(
    view: View,
    savedInstanceState: Bundle?
  ) {
    super.onViewCreated(view, savedInstanceState)

    lockView.onSwipeRefresh { populateList(true) }

    lockView.onFabClicked { checkFabState(true) }

    lockView.onToolbarMenuItemClicked {
      if (it.itemId == R.id.menu_is_system) {
        viewModel.setSystemVisibility(!it.isChecked)
        populateList(true)
      }
    }

    lockView.onListItemClicked { displayLockInfoFragment(it) }

    lockView.onRefreshed { checkFabState(false) }

    visibilityDisposable = viewModel.onSystemVisibilityChanged {
      lockView.onSystemVisibilityChanged(it)
    }

    lockEventDisposable = viewModel.onLockEvent(
        onWhitelist = { populateList(true) },
        onError = { lockView.onModifyEntryError { populateList(true) } }
    )

    databaseChangeDisposable = viewModel.onDatabaseChangeEvent(
        onChange = { lockView.onDatabaseChangeReceived(it.index, it.entry) },
        onError = { ErrorDialog().show(requireActivity(), "db_change_error") }
    )

    fabStateChangeDisposable = viewModel.onFabStateChange {
      onFabStateChanged(it, false)
    }

    val intent = requireActivity().intent
    if (intent.hasExtra(ServiceManager.FORCE_REFRESH_ON_OPEN)) {
      intent.removeExtra(ServiceManager.FORCE_REFRESH_ON_OPEN)

      Timber.d("Launched from notification, force list refresh")
      populateList(true)
    }

    createPinPresenter.bind(viewLifecycleOwner, this)
    confirmPinPresenter.bind(viewLifecycleOwner, this)
  }

  private fun onFabStateChanged(
    state: ServiceState,
    fromClick: Boolean
  ) {
    when (state) {
      ENABLED -> onFabIconLocked(fromClick)
      DISABLED -> onFabIconUnlocked(fromClick)
      PERMISSION -> onFabIconPermissionDenied(fromClick)
      PAUSED -> onFabIconPaused(fromClick)
    }
  }

  private fun populateList(forced: Boolean) {
    populateDisposable = viewModel.populateList(forced,
        onPopulateBegin = { lockView.onListPopulateBegin() },
        onPopulateSuccess = { lockView.onListLoaded(it) },
        onPopulateError = { lockView.onListPopulateError { populateList(true) } },
        onPopulateComplete = { lockView.onListPopulated() }
    )
  }

  override fun onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    lockView.commitListState(outState)
  }

  override fun onStart() {
    super.onStart()
    populateList(false)
    checkFabState(false)
  }

  override fun onResume() {
    super.onResume()
    requireToolbarActivity().withToolbar {
      it.setTitle(R.string.app_name)
      it.setUpEnabled(false)
    }
  }

  override fun onPause() {
    super.onPause()
    lockView.commitListState(null)

    if (isRemoving) {
      toolbarActivity?.withToolbar {
        it.menu.apply {
          removeGroup(R.id.menu_group_list_system)
          removeGroup(R.id.menu_group_list_search)
        }
        it.setOnMenuItemClickListener(null)
      }
    }
  }

  override fun onDestroyView() {
    super.onDestroyView()
    fabStateCheckDisposable.tryDispose()
    populateDisposable.tryDispose()
    fabStateChangeDisposable.tryDispose()
    databaseChangeDisposable.tryDispose()
    lockEventDisposable.tryDispose()
    visibilityDisposable.tryDispose()
  }

  private fun checkFabState(fromClick: Boolean) {
    fabStateCheckDisposable = viewModel.checkFabState {
      onFabStateChanged(it, fromClick)
    }
  }

  private fun displayLockInfoFragment(entry: AppEntry) {
    LockInfoDialog.newInstance(entry)
        .show(requireActivity(), LockInfoDialog.TAG)
  }

  private fun onFabIconLocked(fromClick: Boolean) {
    lockView.onFabIconLocked()

    if (fromClick) {
      if (UsagePermissionChecker.hasPermission(requireContext())) {
        PinConfirmDialog.newInstance(finishOnDismiss = false)
            .show(requireActivity(), PinConfirmDialog.TAG)
      }
    }
  }

  private fun onFabIconUnlocked(fromClick: Boolean) {
    lockView.onFabIconUnlocked()

    if (fromClick) {
      if (UsagePermissionChecker.hasPermission(requireContext())) {
        PinCreateDialog.newInstance()
            .show(requireActivity(), PinCreateDialog.TAG)
      }
    }
  }

  private fun onFabIconPermissionDenied(fromClick: Boolean) {
    lockView.onFabIconPermissionDenied()

    if (fromClick) {
      UsageAccessRequestDialog().show(requireActivity(), "usage_access")
    }
  }

  private fun onFabIconPaused(fromClick: Boolean) {
    lockView.onFabIconPaused()

    if (fromClick) {
      serviceManager.startService(true)
    }
  }

  override fun onConfirmPinBegin() {
  }

  override fun onConfirmPinSuccess(attempt: String) {
    Timber.d("Pin confirm success")
    clearPinPresenter.clear(attempt)
  }

  override fun onConfirmPinFailure(attempt: String) {
    Timber.d("Pin confirm failed")
    lockView.onMasterPinClearFailure()
  }

  override fun onConfirmPinComplete() {
  }

  override fun onCreatePinBegin() {
  }

  override fun onCreatePinSuccess() {
    Timber.d("Pin create success")
    lockView.onMasterPinCreateSuccess()
  }

  override fun onCreatePinFailure() {
    Timber.d("Pin create failure")
    lockView.onMasterPinCreateFailure()
  }

  override fun onCreatePinComplete() {
  }

  override fun onPinClearSuccess() {
    Timber.d("Pin clear success")
    lockView.onMasterPinClearSuccess()
  }

  override fun onPinClearFailed() {
    Timber.d("Pin clear failure")
    lockView.onMasterPinClearFailure()
  }

  companion object {

    const val TAG = "LockListFragment"
  }
}
